cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt

```diff
@@ -0,0 +1,328 @@
+     */
+    private fun evictEntries() {
+        if (!evictsBySize) {
+            return
+        }
+
+        checkNotNull(accessQueue)
+
+        // address any inconsistencies between the queues and map before eviction.
+        cleanUpDeadEntries()
+
+        while (cacheEntries.size > maxSize) {
+            val entryToEvict = accessQueue.first()
+            cacheEntries.remove(entryToEvict.key)
+            writeQueue?.remove(entryToEvict)
+            accessQueue.remove(entryToEvict)
+        }
+    }
+
+    /**
+     * Update the eviction metadata on the [cacheEntry] which has just been read.
+     */
+    private fun recordRead(cacheEntry: CacheEntry<Key, Value>, nowNanos: Long) {
+        if (expiresAfterAccess) {
+            cacheEntry.accessTimeNanos = nowNanos
+        }
+        accessQueue?.add(cacheEntry)
+    }
+
+    /**
+     * Update the eviction metadata on the [CacheEntry] which is about to be written.
+     * Note that a write is also considered an access.
+     */
+    private fun recordWrite(cacheEntry: CacheEntry<Key, Value>, nowNanos: Long) {
+        if (expiresAfterAccess) {
+            cacheEntry.accessTimeNanos = nowNanos
+        }
+        if (expiresAfterWrite) {
+            cacheEntry.writeTimeNanos = nowNanos
+        }
+        accessQueue?.add(cacheEntry)
+        writeQueue?.add(cacheEntry)
+    }
+
+    /**
+     * Although [cacheEntries], [writeQueue], and [accessQueue] are meant to have identical entries,
+     * in theory they can get out of sync with enough concurrency due to thread preemption.
+     * Overtime the inconsistency between the queues and the map might cause memory leaks, e.g.
+     * when performing size-based evictions, an entry in the [cacheEntries] but not in [accessQueue]
+     * would never be evicted.
+     * Calling this function will address these inconsistencies by only retaining the intersection
+     * of [cacheEntries], [writeQueue], and [accessQueue].
+     */
+    private fun cleanUpDeadEntries() {
+        val queues = listOfNotNull(writeQueue, accessQueue)
+        queues.forEach { queue ->
+            val queueSize = queue.size
+            val cacheEntrySize = cacheEntries.size
+
+            if (queueSize < cacheEntrySize) {
+                // remove entries in the map but not in the queue
+                val iterator = cacheEntries.iterator()
+                for (item in iterator) {
+                    val cacheEntry = item.value
+                    if (!queue.contains(cacheEntry)) {
+                        iterator.remove()
+                        // also make sure the other queue does not have it
+                        if (queue == writeQueue) {
+                            accessQueue?.remove(cacheEntry)
+                        } else {
+                            writeQueue?.remove(cacheEntry)
+                        }
+                    }
+                }
+            } else if (queueSize > cacheEntrySize) {
```

Comment:
by having both these checks it implies that you can also have a condition where you have the same number of cache entries in the map and the queue but they do not match.

Assuming that we always record the entry in the queues before adding it to the map (which is the case) we will also actually introduce a bug here where in the following interleaving:
thread 1 records 2 entry in the queue
thread 2 calls cleanup and removes that entry from the queue because it doesn't have a counterpart in the map
thread 1 adds the entry into the map
the next call to cleanup will evict the new entry from the map

Luckily if we always first update the queues then update the map we only need to remove entries in the map that do not have matching entries in the queue (as the only way that they were removed is if the entry was evicted from the cache but was then re-added by the interrupted put thread).

To fix this bug I think we can just remove the `queueSize > cacheEntrySize` branch and make sure to always update the map after the queue(s)
Also, if you are in the process of adding a new entry than it may g